File: src/lib/algorithms/halftone.ts
// Halftone dithering
export function halftoneDithering(
    grayscale: Uint8ClampedArray,
    width: number,
    height: number,
    dotSize: number = 3,
    spacing: number = 5,
    angle: number = 45
  ): ImageData {
    // Create output ImageData
    const outputData = new ImageData(width, height);
    const output = outputData.data;
    
    // Clear the output to white
    for (let i = 0; i < output.length; i += 4) {
      output[i] = 255;     // R
      output[i + 1] = 255; // G
      output[i + 2] = 255; // B
      output[i + 3] = 255; // A
    }
    
    // Convert angle to radians
    const rad = angle * Math.PI / 180;
    const cos = Math.cos(rad);
    const sin = Math.sin(rad);
    
    // Determine the cell size (dot + spacing)
    const cellSize = dotSize + spacing;
    
    // Calculate rotated grid coordinates
    for (let cy = 0; cy < height; cy += cellSize) {
      for (let cx = 0; cx < width; cx += cellSize) {
        // Center of the cell
        const centerX = cx + cellSize / 2;
        const centerY = cy + cellSize / 2;
        
        // Skip if center is outside the image
        if (centerX < 0 || centerX >= width || centerY < 0 || centerY >= height) {
          continue;
        }
        
        // Get the pixel value at the center
        const pixelIdx = Math.floor(centerY) * width + Math.floor(centerX);
        if (pixelIdx >= grayscale.length) continue;
        
        const pixelValue = grayscale[pixelIdx];
        
        // Calculate dot size based on pixel value (darker pixel = larger dot)
        const intensity = 1 - pixelValue / 255;
        const radius = (dotSize / 2) * intensity;
        
        // Draw the dot
        for (let dy = -radius; dy <= radius; dy++) {
          for (let dx = -radius; dx <= radius; dx++) {
            // Rotate the point
            const rotX = dx * cos - dy * sin;
            const rotY = dx * sin + dy * cos;
            
            // Calculate the pixel coordinates
            const x = Math.floor(centerX + rotX);
            const y = Math.floor(centerY + rotY);
            
            // Check if within image bounds
            if (x >= 0 && x < width && y >= 0 && y < height) {
              // Check if within circle
              if (dx * dx + dy * dy <= radius * radius) {
                const idx = (y * width + x) * 4;
                output[idx] = 0;     // R
                output[idx + 1] = 0; // G
                output[idx + 2] = 0; // B
                // Alpha already set to 255
              }
            }
          }
        }
      }
    }
    
    return outputData;
  }

File: rust_wasm/src/halftone.rs
// Halftone dithering algorithm
pub fn halftone_dither(
    input_ptr: *const u8,
    width: usize,
    height: usize,
    dot_size: usize,
    spacing: usize,
    angle: f32,
    output_ptr: *mut u8,
) {
    // Convert the input and output pointers to slices
    let input = unsafe { std::slice::from_raw_parts(input_ptr, width * height) };
    let output = unsafe { std::slice::from_raw_parts_mut(output_ptr, width * height) };
    
    // Clear output buffer
    for i in 0..output.len() {
        output[i] = 0;
    }
    
    // Calculate the rotation matrix for the specified angle
    let angle_rad = angle * std::f32::consts::PI / 180.0;
    let cos_angle = angle_rad.cos();
    let sin_angle = angle_rad.sin();
    
    // Cell size (dot size + spacing)
    let cell_size = dot_size + spacing;
    
    // Process the image in cell-sized blocks
    for y in 0..height {
        for x in 0..width {
            let idx = y * width + x;
            let pixel_value = input[idx];
            
            // Calculate the cell position for this pixel
            let cell_x = (x / cell_size) * cell_size + cell_size / 2;
            let cell_y = (y / cell_size) * cell_size + cell_size / 2;
            
            // Calculate rotated coordinates
            let dx = (x as f32 - cell_x as f32);
            let dy = (y as f32 - cell_y as f32);
            let rotated_x = dx * cos_angle - dy * sin_angle;
            let rotated_y = dx * sin_angle + dy * cos_angle;
            
            // Calculate the max dot radius based on the input value (0-255)
            let max_radius = (dot_size as f32 / 2.0) * (pixel_value as f32 / 255.0);
            
            // Check if this pixel falls within the dot
            let dist = (rotated_x * rotated_x + rotated_y * rotated_y).sqrt();
            
            if dist <= max_radius {
                output[idx] = 255;
            }
        }
    }
} 